### PR TITLE
Monte Carlo Dropout for TCN models

### PR DESCRIPTION
**Pull request by shaginhekvs**
Adds Feature Monte Carlo Dropout for TCN model. 
Example 6 TCN has been updated accordingly to demonstrate Monte Carlo dropout

### CODE REVIEW

**Comment by hrzn** [comment]
There are now conflicts, since I merged the doc-update PR.
I think the conflicts on the notebooks should be easy to fix, if you `git mv` the notebooks on your branch and rename like this:
* `01-darts-intro.ipynb`  --> `00-quickstart.ipynb`
* `04-FFT-examples.ipynb` --> `03-FFT-examples.ipynb`
* `06-TCN-examples.ipynb` --> `05-TCN-examples.ipynb`

File: darts/models/forecasting/tcn_model.py [hunk 1]
```diff
@@ -332,7 +333,11 @@ def __init__(self,
 
         kwargs['input_chunk_length'] = input_chunk_length
         kwargs['output_chunk_length'] = output_chunk_length
-
+        if(dropout_fn is None):
+            dropout_fn = TorchNativeDropout(0.0)
+        if isinstance(dropout_fn, McCompatibleDropout) and (likelihood is not None):
+            logger.warning("Enabling Likelihood and Monte Carlo Dropout makes model doubly stochastic, \
```
Comment (hrzn) [inline darts/models/forecasting/tcn_model.py:339]: Actually I don't think we need this warning - anyway fitting deep architectures to approximate other distributions using stochastic gradient descent with momentum heuristics to approximate negative log likelihoods is also a happy mess :)
IMO it is sufficient to describe precisely in the docstring what setting `enable_mc_dropout=True` in `predict()` does (with a reference to the paper), and maybe say that, if used along likelihoods, it will provide an additional source of stochasticity.

**Review by hrzn — COMMENTED** [review]
Hey @shaginhekvs this is a very nice addition! Very cool to see how easily we can integrate it with our current sampling procedure :)

I think that before merging we can simplify the implementation. For me, the main point would be to let users not worry about building the Dropout objects directly if they don't want to. I.e., let's accept `Union[float, nn.Dropout]` in `predict()` (accepting floats is crucial IMO and also won't break the API), and let's have the "McDropout" object inherit from `nn.Dropout` directly (if doable, unless I miss something) - see my comment in the code. 
That will simplify things quite a bit. Also I think we can remove the super-class `GlobalMCForecastingModel` (but again, maybe I'm missing something).

Another thing before merging: it would be nice to extend it to all models supporting dropout (and maybe add some dropout support here and there to some models that don't have it today).

**Comment by hrzn** [comment]
> There are now conflicts, since I merged the doc-update PR. I think the conflicts on the notebooks should be easy to fix, if you `git mv` the notebooks on your branch and rename like this:
> 
> * `01-darts-intro.ipynb`  --> `00-quickstart.ipynb`
> * `04-FFT-examples.ipynb` --> `03-FFT-examples.ipynb`
> * `06-TCN-examples.ipynb` --> `05-TCN-examples.ipynb`

Actually I think we should keep your changes to TCN notebook only - the other ones are just re-runs and can be stashed.

**Comment by hrzn** [comment]
Another small point: we could rename the boolean `enable_mc_dropout` into `mc_dropout` in the user-facing APIs IMO :)

File: darts/models/forecasting/torch_forecasting_model.py [hunk 2]
```diff
@@ -883,7 +904,7 @@ def _train(self,
             total_loss = 0
 
             for batch_idx, train_batch in enumerate(train_loader):
-                self.model.train()
+                self.__set_train_mode()
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/torch_forecasting_model.py:907]: Do we need `__set_train_mode()`? Since it just calls `model.train()` I believe we can just leave it as before?

Concerning `__set_eval_mode()` we can also customize this  with PyTorch Lightning's `on_predict_model_eval()`

**Comment by hrzn** [comment]
Closing - implemented in https://github.com/unit8co/darts/pull/1013
